services/route_manager/src/main/java/com/futurewei/alcor/route/service/Impl/NeutronRouterServiceImpl.java

```diff
@@ -469,6 +468,197 @@ public ConnectedSubnetsWebResponse getConnectedSubnets(String projectId, String
+            RouteEntry existRoute = (RouteEntry)existRouteEntry.getValue();
+            InternalRoutingRule internalRoutingRule = constructNewInternalRoutingRule(OperationType.DELETE, RoutingRuleType.DEFAULT, existRoute, null);
+
+            updateRoutes.add(internalRoutingRule);
+
+        }
+
+        // construct List<InternalSubnetRoutingTable>
+        List<InternalSubnetRoutingTable> subnetRoutingTables = new ArrayList<>();
+
+        InternalSubnetRoutingTable internalSubnetRoutingTable = new InternalSubnetRoutingTable();
+        if (owner != null) {
+            internalSubnetRoutingTable.setSubnetId(owner);
+        } else {
+            internalSubnetRoutingTable.setSubnetId(existRouteTable.getOwner());
+        }
+        internalSubnetRoutingTable.setRoutingRules(updateRoutes);
+        subnetRoutingTables.add(internalSubnetRoutingTable);
+
+        // construct UpdateRoutingRuleResponse
+        UpdateRoutingRuleResponse updateRoutingRuleResponse = new UpdateRoutingRuleResponse(subnetRoutingTables, hostRouteToSubnet);
+
+        return updateRoutingRuleResponse;
+    }
+
+    @Override
+    public InternalRouterInfo constructInternalRouterInfo (List<InternalSubnetRoutingTable> internalSubnetRoutingTableList) {
+
+        String requestId = UUID.randomUUID().toString();
+        InternalRouterInfo internalRouterInfo = new InternalRouterInfo();
+
+        InternalRouterConfiguration internalRouterConfiguration = new InternalRouterConfiguration();
+
+        internalRouterConfiguration.setSubnetRoutingTables(internalSubnetRoutingTableList);
+
+        internalRouterConfiguration.setRevisionNumber(ConstantsConfig.REVISION_NUMBER);
+        internalRouterConfiguration.setFormatVersion(ConstantsConfig.FORMAT_VERSION);
+        internalRouterConfiguration.setRequestId(requestId);
+        internalRouterConfiguration.setHostDvrMac("");
+        internalRouterConfiguration.setMessageType(MessageType.FULL);
+
+        internalRouterInfo.setOperationType(OperationType.INFO);
+        internalRouterInfo.setRouterConfiguration(internalRouterConfiguration);
+
+        return internalRouterInfo;
+
+    }
+
+    private InternalRoutingRule constructNewInternalRoutingRule(OperationType operationType, RoutingRuleType routingRuleType, RouteEntry route, NewRoutesRequest newRouteRequest) {
+        if (route == null && newRouteRequest == null) {
+            return new InternalRoutingRule();
+        }
+
+        InternalRoutingRule internalRoutingRule = new InternalRoutingRule();
+
+        InternalRoutingRuleExtraInfo routingRuleExtraInfo = new InternalRoutingRuleExtraInfo();
+        //routingRuleExtraInfo.setNextHopMac();
+        // TODO: insert destination type - if it is vpc router, configure value according to target
+        routingRuleExtraInfo.setDestinationType(VpcRouteTarget.LOCAL);
+
+        if (route == null) {
+            internalRoutingRule.setId(route.getId());
```

Comment:
Is that possible for `route == null`, that's mean existing route doesn't exist?
If it's `null`, how can you `getId()` from a `null`?

Reply:
Yes, route == null is possible I think. And the two lines of code here are indeed in reverse order